src/Reflex/Dom/ACE.hs

```diff
@@ -143,233 +142,155 @@ data AceDynConfig = AceDynConfig
-          jscb <- asyncCallback1 $ \_ -> liftIO $ do
-              v <- getValueACE ace
-              cb v
-          js_setupValueListener ace jscb
-    performEventAsync (act <$ pb)
+setValueACE :: MonadJSM m => Text -> AceInstance -> m ()
+setValueACE t (AceInstance ace) =
+  liftJSM $ void $ ace ^. js2 "setValue" t (-1 :: Int)
 
-foreign import javascript unsafe
-  "(function(){ $1['on']('change', $2); })()"
-  js_setupValueListener :: AceRef -> Callback (JSVal -> IO ()) -> IO ()
-#else
-setupValueListener = error "setupValueListener: can only be used with GHCJS"
-#endif
+
+------------------------------------------------------------------------------
+setupValueListener :: MonadWidget t m => AceInstance -> m (Event t Text)
+setupValueListener (AceInstance ace) = do
+  pb  <- getPostBuild
+  let act cb = liftJSM $ do
+        jscb <- asyncFunction $ \_ _ _ -> getValueACE (AceInstance ace) >>= liftIO . cb
+        void $ ace ^. js2 "on" "change" jscb
+  performEventAsync (act <$ pb)
 
 
 ------------------------------------------------------------------------------
 -- | Main entry point
 aceWidget
-    :: MonadWidget t m
+    :: (MonadWidget t m, MonadJSM m)
     => AceConfig
     -> AceDynConfig
     -> Event t AceDynConfig
     -> Text
     -> m (ACE t)
 aceWidget ac adc adcUps initContents = do
     attrs <- holdDyn (addThemeAttr adc) (addThemeAttr <$> adcUps)
-    aceDiv <- fmap fst $ elDynAttr' "div" attrs $ text initContents
-#ifdef ghcjs_HOST_OS
+    aceDiv <- fmap fst $ elDynAttr' (T.pack "div") attrs $ text initContents
     let aceEl = _element_raw aceDiv
-#else
-    let aceEl = (error "aceWidget is only available to ghcjs") aceDiv
-#endif
     pb <- getPostBuild
-    aceUpdates <- performEvent (liftIO (startACE aceEl ac) <$ pb)
+    aceUpdates <- performEvent (liftJSM (startACE aceEl ac) <$ pb)
 
     res <- widgetHold (return never) $ setupValueListener <$> aceUpdates
     aceDyn <- holdDyn Nothing $ Just <$> aceUpdates
     updatesDyn <- holdDyn initContents $ switchPromptlyDyn res
 
     let ace = ACE aceDyn updatesDyn
-    withAceRef ace (setThemeACE (_aceDynConfigTheme adc) <$ pb)
-    withAceRef ace (setThemeACE . _aceDynConfigTheme <$> adcUps)
+    withAceInstance ace (setThemeACE (_aceDynConfigTheme adc) <$ pb)
+    withAceInstance ace (setThemeACE . _aceDynConfigTheme <$> adcUps)
     return ace
   where
     static = _aceConfigElemAttrs ac
-    themeAttr t = " ace-" <> T.pack (show t)
+    themeAttr t = T.pack $ " ace-" <> show t
     addThemeAttr c = maybe static
-      (\t -> M.insertWith (<>) "class" (themeAttr t) static)
+      (\t -> M.insertWith (<>) (T.pack "class") (themeAttr t) static)
       (_aceDynConfigTheme c)
 
 
 ------------------------------------------------------------------------------
--- | Convenient helper function for running functions that need an AceRef.
-withAceRef
+-- | Convenient helper function for running functions that need an AceInstance.
+withAceInstance, withAceRef
```

Comment:
If we're going with the new name `AceInstance` we might as well kill the old function since it's not a breaking change.  If we're going with the old name, then we don't need the new one.